### ISSUE TITLE
primary mac auto registration

### ISSUE THREAD

**Issue by pcnate** [FOGProject/fog-client#91]
When registering hosts automatically with the fog client, if multiple network adapters are present, the primary isn't necessarily the wired ethernet port. Booting to fog results in client not registered since the additional mac addresses are not considered. Tried updating the mac address through the web UI but it does not save these changes.

I would suggest three things:
1) There is a bug on the host general tab where saving mac addresses does not work
2) Additional mac addresses should be checked for when checking for host registration. Unless I am wrong, you cannot boot to USB Ethernet or WiFi adapters so maybe they should be separated from the on-board mac addresses.
3) primary mac address should be selected better. Many machines have dual wired Ethernet, so maybe the option to have 2-4.

**Comment by Sebastian-Roth** [comment]
There are a lot of things in this one post and I won't get to provide an answer for all of those just now. So let's start with two simple things at first. You are saying that there is a bug in the general tab of the web interface when saving MAC addresses. We need to know which version of FOG you are running to be able to further look into this. You might better post this part of the issue in https://github.com/FOGProject/fogproject or https://forums.fogproject.org as this has nothing to do with the client.

Second, we are working on getting rid of the "primary" MAC address being the key to identify hosts altogether. We'll be using the system UUID to identify hosts then. As far as I can see this should completely solve your issue. It's not too big of a change but we don't have lots of spare time as of right now. So it might still take a little bit of time till be get this one done.

Please let me know if you think this change is not going to solve what you requested.

**Comment by pcnate** [comment]
I forgot I was posting this under the client section. If you are removing the mac address identification, should I even worry about posting this issue under the fogproject repo?

I've encountered this issue on a single site running 1.4.2 and later 1.4.4 on Debian 9. I've had to resort to not deploying the fog client remotely for laptops or wireless machines so then I must plug them into Ethernet and run the full registration which is only a bit inconvenient. 

Will the UUID identify a system without an OS installed? If so, then it should completely solve my problem. I'd prefer the hardware to be identified and the MAC does this. If not, can this be a setting?

Are these changes expected under 1.5.x or a later major version?

**Comment by Sebastian-Roth** [comment]
> I forgot I was posting this under the client section. If you are removing the mac address identification, should I even worry about posting this issue under the fogproject repo?
Are you able to replicate this issue in the web UI (version 1.4.4)? I am wondering if this is somehow caused by a hickup in your database and not a general issue?! Maybe try this: https://wiki.fogproject.org/wiki/index.php/Troubleshoot_MySQL#Database_Maintenance_Commands

> Will the UUID identify a system without an OS installed?
Yes it will. This UUID or sometimes also called GUID is read from SMBIOS and does not need an OS on disk.

> I'd prefer the hardware to be identified and the MAC does this.
As we see the MAC does **not** really identify a system. This is why we want to switch over to using the UUID/GUID... Or maybe I don't get what you mean by that?!

> Are these changes expected under 1.5.x or a later major version?
Definitely. But there is no time schedule on when this will be released.

**Comment by Sebastian-Roth** [comment]
Sorry for the delay but I had to get my new building environment ready. Got it mostly up now and started to work on adding the system UUID as primary identifier in the fog-client. On the server side Tom has [added the code](https://github.com/Sebastian-Roth/fogproject/commit/245460bfef5cce89f29f494d514a2cdef251e122) already. For that to use you need to be on FOG 1.5.0 RC-1 at least!

**Comment by pcnate** [comment]
Is the uuid going to have issues with upgrading? I don't have a lot of time
to spend on fixing things. I could prolly do it off company time. Not every
machine has had a hardware inventory. That would be great to be possible to
get from the client for laptops.

**Comment by mastacontrola** [comment]
The UUID will not have issues with upgrading.  The new fog client auto updates. The server side defaults back to the legacy method of host detection (using the mac address of the host) if the system UUID is not already defined.  When the device pxe boots next, however, it will update the inventory record with it's relevant UUID.

**Comment by pcnate** [comment]
Ok. So laptops need a hard wired connections again after the upgrade

**Comment by mastacontrola** [comment]
Not necessarily. My hope would be, that the client code being added would also have the ability to update the host's UUID for us if we find a matching, non-set UUID host based on the MAC Address.  It will take a little coordination of course to ensure the client and server code mesh properly, but it should not be difficult. Ultimately, if the mac address is registered to the host, it should be able to update the UUID using the fog client as well as it is able to do so through pxe booting/inventory updates.
